tests/fixtures/optimizer/qualify_columns.sql

```diff
@@ -0,0 +1,124 @@
+--------------------------------------
+-- Qualify columns
+--------------------------------------
+SELECT a FROM x;
+SELECT x.a AS a FROM x;
+
+SELECT a AS a FROM x;
+SELECT x.a AS a FROM x;
+
+SELECT x.a FROM x;
+SELECT x.a AS a FROM x;
+
+SELECT x.a AS a FROM x;
+SELECT x.a AS a FROM x;
+
+SELECT a AS b FROM x;
+SELECT x.a AS b FROM x;
+
+SELECT 1, 2 FROM x;
+SELECT 1 AS "_col0", 2 AS "_col1" FROM x;
+
+--------------------------------------
+-- Derived tables
+--------------------------------------
+SELECT y.a AS a FROM (SELECT x.a AS a FROM x) AS y(a);
+SELECT y.a AS a FROM (SELECT x.a AS a FROM x) AS y(a);
+
+SELECT y.a FROM (SELECT a AS a FROM x) AS y(a);
+SELECT y.a AS a FROM (SELECT x.a AS a FROM x) AS y(a);
+
+SELECT y.a FROM (SELECT x.a AS a FROM x) AS y;
+SELECT y.a AS a FROM (SELECT x.a AS a FROM x) AS y(a);
```

Comment:
why is it as y(a) ? this doesn’t seem to be standard syntax is it? dan we just leave it AS y

Reply:
I _think_ the alias column list is standard... 
Here it is in MySQL: https://dev.mysql.com/doc/refman/8.0/en/join.html
Here it is in Presto: https://prestodb.io/docs/current/sql/select.html

It can potentially help with some later steps. Take this for example:

```sql
SELECT 
    *
FROM (
    SELECT a, b FROM x 
    UNION 
    SELECT a, b FROM y) AS z(a, b)
```

The alias column list is sort of like a cache of the output columns...
```python
subquery = expression.find(Subquery)
name = subquery.args["alias"].text("this")
outputs = subquery.args["alias"].args["columns"]

outputs_to_keep = outputs.intersection(columns_by_selectable[name])
projection_pushdown(subquery, outputs_to_keep)
```

Now, this is probably the equivalent of the "named_selections" method @nolamesa added. But something feels nice and clean about making the AST more explicit... it's nice to have a predictable structure. 🤷  I could go either way.

Reply:
i see what you're saying, but the issue with alias columns is i feel it's pretty rare and just duplicate of the aliases / names providied in SELECTS, so named_selections provides exactly this.

since we already force alias everything explicitly, i'd say we should get rid of the table column aliases. if the user provides them, we should "push those down" so that they don't exist in the canonical form.

Reply:
what i  mean is if a user writes (select a AS b) x (c), we should optimize this to be (SELECT a AS c) i think this is safe? if for some reason this is not safe, then ya maybe we need to do this table column alias

Reply:
```sql
select 1 AS a
where a > 1
```

for example, this is not legal, meaning the aliases don't really have scope inside and mostly serve for projection

and so 
```sql
(select 1 AS a
... reference a
) AS table(c)
```

transformed to

```sql
(select 1 AS c
... reference a
) AS table

shouldn't happen, but again, i've never seen a sql query using alias projection EXCEPT in cases of explosive queries

Reply:
Ok, I'm on board. It'll be more context we have to push down into subqueries... but I think that's fine.